Use config for the SensesField

Larkinabout opened the pull request:
Intended to allow custom senses by editing `CONFIG.DND5E.senses`.

![image](https://github.com/user-attachments/assets/90e809c7-75d5-4b9e-9d42-8ac941c8b98c)

Larkinabout committed:
Use config for the SensesField

krbz999 commented on module/data/shared/senses-field.mjs:
```diff
@@ -6,11 +6,12 @@ const { NumberField, StringField } = foundry.data.fields;
 export default class SensesField extends foundry.data.fields.SchemaField {
   constructor(fields={}, options={}) {
     const numberConfig = { required: true, nullable: true, integer: true, min: 0, initial: null };
+    const senses = Object.entries(CONFIG.DND5E.senses).reduce((acc, [k, label]) => {
```
This should probably be a `MappingField` if it should be configurable. Otherwise data is lost as soon as a module is disabled, or if a module removes any of the default senses.

Larkinabout replied:
I did do some tests and it seemed to retain the values set against an actor for the core senses, and even custom senses, when removed and re-added. But, I'll admit I don't have much comprehension of it beyond where it's defining the senses initially. I'll see if I can figure out what a `MappingField` will do.

Larkinabout replied:
Struggling a bit to understand how `MappingField` is intended to work in this context, with `units` and `special` being handled alongside the config properties. Creating a `MappingField` within fields for the config properties makes sense, as those are actual senses compared to units and special. However, it looks like it'll complicate handling elsewhere (which might be necessary).

That said, I may be misunderstanding the suggested application.

Is this what you were envisioning? Or was your suggestion something else?

![image](https://github.com/user-attachments/assets/a6e4bc9a-9e83-4c46-b37b-37e74faf78f5)

arbron replied:
The way I handle it in Black Flag is have something like:

```javascript
{
  senses: {
    types: MappingField,
    special: StringField,
    units: StringField
  }
}
```

This will of course require a migration and changes anywhere in the system where it accesses the sense values (part of the reason I haven't gotten around to it yet, big project).

Larkinabout replied:
At least I wasn't totally off the mark! I'll see if I can slowly chip away at it.

Larkinabout replied:
With this part: https://github.com/foundryvtt/dnd5e/blob/99000e24daf04d0594f9a65c4c2a73f3f398dcae/module/applications/shared/movement-senses-config.mjs#L82

How would I build it out using the `MappingField`' so that the `NumberField`s are provided to the template with `name` and `label`? Is there an existing function to do it? Otherwise, what's the correct way to provide a mapping field for the template?

Additionally, I'm guessing it'd be better to convert movement in the same way at the same time as senses?

Larkinabout committed:
Add types to senses object

Larkinabout committed:
Merge branch '4.2.x' into use-config-for-senses-field

Larkinabout committed:
Merge branch '4.2.x' into use-config-for-senses-field

Larkinabout committed:
Add types to movement object, add migration
- Add the types object to the movement object.
- Add migration to migrate movements and senses to the types object.

Larkinabout commented:
I believe the only thing left to do is resolve `MovementSensesConfig._preparePartContext`, though I'm still not clear on how to do it. I'll keep trying to figure it out, but would be grateful for anyone's insight to speed me along.

Larkinabout committed:
Add human-readable attributes for movement and senses

Larkinabout committed:
Update movement/senses config

Larkinabout commented:
Note to remind that we may wish to also move `group` actor movement types into a `types` object for consistency, but for the moment I've handled it in `movement-senses-config.mjs`.

Also, would be useful to allow `group` movement to also be configurable. `CONFIG.DND5E.vehicleTypes` could be viable, but it includes `space`, which would need to be excluded by default. This would cause an issue if a user actually wants to add space as a movement type.

arbron commented:
@Larkinabout It would be good to take the compendium changes out of this PR, we can handle them later and we'll want to make sure that the migrations in place are enough to keep compendium content working even if it doesn't have the new data layout (plus it makes it easier to review)

Larkinabout committed:
Revert packs

Larkinabout committed:
Add data migrations for attributes and race

krbz999 commented:
This probably also needs migrations for ActiveEffects. All AEs that modify `system.attributes.senses.*` should point at something different, or we just support the same deprecated property but point it elsewhere in the AE document class.

Larkinabout committed:
Migrate effects, Shim old movement/senses

Larkinabout committed:
Fix NaN movement for vehicles

Larkinabout committed:
Small refactor for consistency

Larkinabout committed:
Merge branch '4.2.x' into use-config-for-senses-field

Larkinabout committed:
Merge branch '4.3.x' into use-config-for-senses-field

Larkinabout committed:
Merge branch '4.3.x' into use-config-for-senses-field

arbron commented on module/migration.mjs:
```diff
@@ -843,6 +848,30 @@ function _migrateEffectArmorClass(effect, updateData) {
+function _migrateEffectMovementSenses(effect, updateData) {
+  let containsUpdates = false;
+  const changes = (effect.changes || []).map(c => {
+    if ((c.key.startsWith("system.attributes.movement.") || c.key.startsWith("system.attributes.senses."))
```
This looks like it will move any changes targeting `attributes.movement.units`, `attributes.movement.hover`, or `attributes.senses.special` into the types object.

arbron commented on module/applications/actor/base-sheet.mjs:
```diff
@@ -249,15 +249,12 @@ export default class ActorSheet5e extends ActorSheetMixin(ActorSheet) {
-    if ( largestPrimary ) {
-      speeds.push([movement.walk, `${game.i18n.localize("DND5E.MovementWalk")} ${movement.walk}`]);
-    }
+    let speeds = Object.entries(CONFIG.DND5E.movementTypes).filter(([k]) => !(!largestPrimary && k === "walk"))
```
```suggestion
    let speeds = Object.entries(CONFIG.DND5E.movementTypes)
      .filter(([k]) => largestPrimary || (k !== "walk"))
```

arbron commented on module/applications/shared/movement-senses-config.mjs:
```diff
@@ -74,15 +78,20 @@ export default class MovementSensesConfig extends BaseConfigSheet {
-      value: context.data[key],
-      placeholder: placeholderData?.[key] ?? ""
-    }));
+    context.types = this.types.map(([k, label]) => {
+      return {
+        field: isGroup ? context.fields[k] : context.fields.types.model,
+        key: `system.${this.keyPath}.${isGroup ? "" : "types."}${k}`,
+        label,
+        placeholder: (isGroup ? placeholderData?.[k] : placeholderData?.types?.[k]) ?? "",
+        value: isGroup ? context.data[k] : context.data.types?.[k]
+      };
+    });
```
```suggestion
    context.types = this.types.map(([k, label]) => ({
      field: isGroup ? context.fields[k] : context.fields.types.model,
      key: `system.${this.keyPath}.${isGroup ? "" : "types."}${k}`,
      label,
      placeholder: (isGroup ? placeholderData?.[k] : placeholderData?.types?.[k]) ?? "",
      value: isGroup ? context.data[k] : context.data.types?.[k]
    }));
```

Larkinabout committed:
Resolve quick fixes

Larkinabout committed:
Migrate only keys in config

Larkinabout committed:
Merge branch '4.3.x' into use-config-for-senses-field

arbron commented on module/data/actor/templates/attributes.mjs:
```diff
@@ -121,6 +131,29 @@ export default class AttributesFields {
+    ];
+
+    for (const { key, config } of movementSenses) {
+      const attr = source[key];
```
Because it is called `AttributesField._migrate(source.attributes)`, and if you call `actor.update({ "details.cr": 5 });` then `source` will just equal `{ details: { cr: 5 } }`, so migrate will get `undefined`.

Returning early would probably be okay (in which case you could adjust the JSDocs back because `source` would always exist).

Larkinabout committed:
Add early return on _migrate

Larkinabout committed:
Merge branch 'use-config-for-senses-field' of https://github.com/Larkinabout/dnd5e into use-config-for-senses-field

arbron commented:
@Larkinabout Please give a comment over on https://github.com/foundryvtt/dnd5e/issues/4667 so I can assign it to you

Fyorl commented:
This work should be done with the vision/detection mode work in mind and in concert with it.
